### PR TITLE
Associate Apollo context with `React.createContext` (instead of using a local `WeakMap`) again

### LINKED ISSUE #8790: Multiple apollo-client instances will not reuse the same Apollo Context in React
We are using a Microfrontend(MFE) architecture, where Apollo client is packaged individually with each MFE. Until 3.2.0 Apollo context was created and stored on the `React` module object on a symbol, but the behavior was changed with this change https://github.com/apollographql/apollo-client/commit/c46e02f580ea5bf4b7f7367f8bf23123a230827a

**Intended outcome:**

Multiple `@apollo/client` modules must reuse the same React context to consume from the provider when using `useApolloClient` or other hooks.

**Actual outcome:**

Since Apollo now creates a `WeakMap` in the module scope of `@apollo/client` this weak map is recreated for each individual module instances, the earlier approach was caching it on the `React` object which worked. I do not see how creating this `Weakmap` helps compared to just creating the React context as a module scoped singleton variable.

Code here: https://github.com/apollographql/apollo-client/blob/main/src/react/context/ApolloContext.ts#L17

image

**How to reproduce the issue:**

- Package multiple copies of `@apollo/client` in an app. Create a `ApolloContext` Provider using one of the copies, and provide a `client` instance.
- Try using `useApolloClient` hook from another copy, to see that the `client` is undefined.

**Versions**

3.4.0

### PR DESCRIPTION
**Pull request by benjamn**
Should fix #8790, by allowing multiple copies of `@apollo/client` (uncommon) to share the same `React.createContext[contextKey]` object.

A lot has changed since #7371, but I am not worried about `React.createContext` being undefined, because (in this module, `ApolloContext.ts`), the `React` object is now a namespace object (thanks to `import * as React from 'react'`), and we call `React.createContext` unconditionally within `getApolloContext` already, so `React.createContext` must be defined when `getApolloContext` is called, or this code would be totally broken.

This implementation uses `React.createContext[contextKey]` instead of `React[contextKey]` (as in #7371) because the `React` namespace object is likely frozen/non-extensible, whereas `React.createContext` is an ordinary function object.

### CODE REVIEW

**Review by hwillson — APPROVED** [review]
Looks great @benjamn!

**Comment by ruben-nogueira** [comment]
@benjamn My vue 2.6 projects are failing to run/build after upgrading the apollo client dependency to 3.4.12.

Log:
```
Step #2: > xxxxxxxxxxxx@0.1.0 build /workspace
Step #2: > vue-cli-service build
Step #2: 
Step #2:  ERROR  Error loading /workspace/vue.config.js:
Step #2:  ERROR  Error: Cannot find module 'react'
Step #2: Require stack:
Step #2: - /workspace/node_modules/@apollo/client/react/context/context.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/react/react.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/main.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/utilities/utilities.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/link/core/core.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/core/core.cjs.js
Step #2: - /workspace/xxxxxxxx.js
Step #2: - /workspace/yyyyy.js
Step #2: - /workspace/zzzzzzz.js
Step #2: - /workspace/vue.config.js
Step #2: - /workspace/node_modules/@vue/cli-shared-utils/lib/module.js
Step #2: - /workspace/node_modules/@vue/cli-shared-utils/index.js
Step #2: - /workspace/node_modules/@vue/cli-service/bin/vue-cli-service.js
Step #2: Error: Cannot find module 'react'
Step #2: Require stack:
Step #2: - /workspace/node_modules/@apollo/client/react/context/context.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/react/react.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/main.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/utilities/utilities.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/link/core/core.cjs.js
Step #2: - /workspace/node_modules/@apollo/client/core/core.cjs.js
Step #2: - /workspace/xxxxx.js
Step #2: - /workspace/yyyyy.js
Step #2: - /workspace/zzzzzzzzz.js
Step #2: - /workspace/vue.config.js
Step #2: - /workspace/node_modules/@vue/cli-shared-utils/lib/module.js
Step #2: - /workspace/node_modules/@vue/cli-shared-utils/index.js
Step #2: - /workspace/node_modules/@vue/cli-service/bin/vue-cli-service.js
Step #2:     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:893:15)
Step #2:     at Function.Module._load (internal/modules/cjs/loader.js:743:27)
Step #2:     at Module.require (internal/modules/cjs/loader.js:965:19)
Step #2:     at require (internal/modules/cjs/helpers.js:88:18)
Step #2:     at Object.<anonymous> (/workspace/node_modules/@apollo/client/react/context/context.cjs.js:6:13)
Step #2:     at Module._compile (internal/modules/cjs/loader.js:1076:30)
Step #2:     at Object.Module._extensions..js (internal/modules/cjs/loader.js:1097:10)
Step #2:     at Module.load (internal/modules/cjs/loader.js:941:32)
Step #2:     at Function.Module._load (internal/modules/cjs/loader.js:782:14)
Step #2:     at Module.require (internal/modules/cjs/loader.js:965:19)
```
Reverting to 3.4.11 fixes the problem. Thanks.

File: src/utilities/observables/subclassing.ts [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import { Observable } from "./Observable";
+import { canUseSymbol } from "..";
```
Comment (benjamn) [inline src/utilities/observables/subclassing.ts:2]: This line seems to have caused a number of problems, hopefully all fixed by #8817.

**Comment by benjamn** [comment]
@ruben-nogueira Please try `@apollo/client@3.4.13` (which includes #8817) when you have a chance!

**Comment by ruben-nogueira** [comment]
@benjamn it's okay. Thanks again. :+1:

**Pull request merged**